src/evict/evict_lru.c

```diff
@@ -1987,6 +1995,233 @@ __evict_walk_tree(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, u_int max_ent
         break;
     }
 
+    return (ret);
+}
+
+/*
+ * __evict_should_give_up_walk --
+ *     Check if we should give up on the current walk.
+ */
+static WT_INLINE bool
+__evict_should_give_up_walk(WT_SESSION_IMPL *session, uint64_t pages_seen, uint64_t pages_queued,
+  uint64_t min_pages, uint32_t target_pages)
+{
+    WT_BTREE *btree;
+    bool give_up;
+
+    btree = S2BT(session);
+
+    /*
+     * Check whether we're finding a good ratio of candidates vs pages seen. Some workloads create
+     * "deserts" in trees where no good eviction candidates can be found. Abandon the walk if we get
+     * into that situation.
+     */
+    give_up = !__wt_cache_aggressive(session) && !WT_IS_HS(btree->dhandle) &&
+      pages_seen > min_pages &&
+      (pages_queued == 0 || (pages_seen / pages_queued) > (min_pages / target_pages));
+    if (give_up) {
+        /*
+         * Try a different walk start point next time if a walk gave up.
+         */
+        switch (btree->evict_start_type) {
+        case WT_EVICT_WALK_NEXT:
+            btree->evict_start_type = WT_EVICT_WALK_PREV;
+            break;
+        case WT_EVICT_WALK_PREV:
+            btree->evict_start_type = WT_EVICT_WALK_RAND_PREV;
+            break;
+        case WT_EVICT_WALK_RAND_PREV:
+            btree->evict_start_type = WT_EVICT_WALK_RAND_NEXT;
+            break;
+        case WT_EVICT_WALK_RAND_NEXT:
+            btree->evict_start_type = WT_EVICT_WALK_NEXT;
+            break;
+        }
+
+        /*
+         * We differentiate the reasons we gave up on this walk and increment the stats accordingly.
+         */
+        if (pages_queued == 0) {
+            WT_STAT_CONN_INCR(session, cache_eviction_walks_gave_up_no_targets);
+            WT_STAT_DSRC_INCR(session, cache_eviction_walks_gave_up_no_targets);
+        } else {
+            WT_STAT_CONN_INCR(session, cache_eviction_walks_gave_up_ratio);
+            WT_STAT_DSRC_INCR(session, cache_eviction_walks_gave_up_ratio);
+        }
+    }
+
+    return (give_up);
+}
+
+/*
+ * __evict_queue_page --
+ *     Check if we should queue the page for eviction. Queue it to the urgent queue or the regular
+ *     queue.
+ */
+static WT_INLINE void
+__evict_queue_page(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, WT_REF *ref,
```

Comment:
nit: Queuing can fail:
```suggestion
__evict_try_queue_page(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue, WT_REF *ref,
```